lib/active_storage/service/azure_service.rb

```diff
@@ -0,0 +1,69 @@
+require 'azure/storage'
+require "active_support/core_ext/numeric/bytes"
+
+class ActiveStorage::Service::AzureService < ActiveStorage::Service
+  attr_reader :client, :path, :blobs, :container
+
+  def initialize(path:, storage_account_name:, storage_access_key:, container:)
+    @client = Azure::Storage::Client.create(storage_account_name: storage_account_name, storage_access_key: storage_access_key)
+    @blobs = client.blob_client
+    @container = container
+    @path = path
+  end
+
+  def upload(key, io, checksum: nil)
+    instrument :upload, key, checksum: checksum do
+      begin
+        blobs.create_block_blob(container, key, io, content_md5: checksum)
+      rescue Azure::Core::Http::HTTPError
+        raise ActiveStorage::IntegrityError
+      end
+    end
+  end
+
+  def download(key)
+    instrument :download, key do
+      _, io = blobs.get_blob(container, key)
+      io
+    end
+  end
+
+  def delete(key)
+    instrument :delete, key do
+      blobs.delete_blob(container, key)
+    end
+  end
+
+  def exist?(key)
+    instrument :exist, key do |payload|
+      answer = blob_for(key).present?
+      payload[:exist] = answer
+      answer
+    end
+  end
+
+  def url(key, expires_in:, disposition:, filename:)
+    instrument :url, key do |payload|
+      blob = blob_for(key)
+      generated_url = url_for(blob)
+
+      payload[:url] = generated_url
+
+      generated_url
+    end
+  end
+
+  def url_for_direct_upload(key, expires_in:, content_type:, content_length:)
+    raise NotImplemented
+  end
+
+  private
+
+  def url_for(blob)
+    "#{path}/#{container}/#{blob.name}"
+  end
+
+  def blob_for(key)
+    blobs.get_blob_properties(container, key)
+  end
```

Comment:
Style: indent an extra level under private and leave no blank line between the access modifier and the first method under it.